037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+   * the `spec.autoApproval` from the template as described by this [proposal](https://github.com/strimzi/proposals/pull/56).
+4. The `KafkaRebalanceAssemblyOperator` detects the new rebalancing resource and acts as usual by requesting Cruise Control to generate an optimization proposal via the REST API on the `/add_broker` endpoint.
+5. When the optimization proposal is ready, the `KafkaRebalanceAssemblyOperator` checks the auto-approval:
+   * If criteria are met or they are not specified at all, it approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
+   * If criteria are not met, the `KafkaRebalance` custom resource status will move to `NotReady`, using a warning condition to describe why it was not approved automatically.
+6. If/When the rebalancing ends successfully, the corresponding `KafkaRebalance` custom resource is deleted by the `KafkaReconciler`.
+
+#### Error handling
+
+During the above flow, errors can happen after the scaling up:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state.
+The cluster operator logs a warning about the rebalancing not happened.
+It's the user who should deal with this scenario:
+
+* Checking the reason why the `KafkaRebalance` is in the `NotReady` state and deleting the custom resource.
+* If the user still wants rebalancing, they have to create a new `KafkaRebalance` custom resource manually.
+
+### Auto-rebalancing on scaling down
+
+1. The user scales down the cluster by decreasing the `spec.kafka.replicas` field.
+2. The `KafkaReconciler` takes care of it and, if the `spec.kafka.autoRebalance` contains an entry with `remove-brokers` mode, it creates a `KafkaRebalance` custom resource named as `auto-<cluster-name>-rebalance` by using the corresponding template and sets:
+   * the `spec.mode` as `remove-brokers` and the `spec.brokers` with the list of brokers that will be removed.
+   * the `spec.autoApproval` from the template as described by this [proposal](https://github.com/strimzi/proposals/pull/56).
+3. The `KafkaRebalanceAssemblyOperator` detects the new rebalancing resource and acts as usual by requesting Cruise Control to generate an optimization proposal via the REST API on the `/remove_broker` endpoint.
+4. When the optimization proposal is ready, the `KafkaRebalanceAssemblyOperator` checks the auto-approval:
+   * If criteria are met or they are not specified at all, it approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
+   * If criteria are not met, the `KafkaRebalance` custom resource status will move to `NotReady`, using a warning condition to describe why it was not approved automatically.
```

Comment:
As mentioned above, I do not think this should have any criteria. The only failure should be when the rebalance cannot be done because of e.g. lack of resources etc.

Reply:
Having criteria is anyway related to how the https://github.com/strimzi/proposals/pull/56 will move forward but in any case we can decide to not having criteria for the auto-rebalancing. But why do you think so?
If we think about some feasible criteria in https://github.com/strimzi/proposals/pull/56 (i.e. max MB to move, max replicas to move, ...) why do you think criteria should not be applied for auto rebalancing?

Reply:
Criteria were removed by the auto-approval proposal.